usr/new/notes/man/newsoutput.8
.TH NEWSOUTPUT 8 "University of Illinois"
.SH NAME
newsoutput, newsinput - notes/news gateway programs
.SH SYNOPSIS
.B newsoutput
[
.B "-a"
or 
.BI "-s" sysname
or
.B "-c"
.I "filename"
] [
.B "-b"
] [
.B "-v"
] [
.B "-o"
.I date-spec
] [
.B "-f"
file
]
topic [ ... ]
.PP
.B newsinput
[
.B "-t"
] [
.B "-1"
]
.SH DESCRIPTION
.I Newsoutput
and
.I newsinput
manage the transfer of articles between a notesfile system
and a news system.

.I Newsoutput
gateways notesfile-generated articles to the news system.
Several options determine the types of articles to be
transfered.
The
.B -a
option specifies that any notesfile-generated article that
has not been in the news system before is to be sent to 
news.
The
.BI "-s" "sysname"
option specifies that only articles from 
.I sysname
are to be sent to the news system.
The
.B "-c"
option directs
.I newsoutput
to gateway articles written at any of the systems whose
names appear in the file
.IB filename "."
If none of these options is specified, only locally
generated articles are sent to the news.
In normal operation, the
.B -a
or
.B -c
options are most typically used;
.B -s
is included for backwards compatibility.
.PP
The 
.B -b
option specifies that newsoutput should generate articles
which will be parsed correctly by previous versions of the
notesfile/news gateway code.
Files containing lists of notesfiles can 
be specified through the 
.B -f
option.
Use
.B -v
for a more verbose output than normal;
default operation produces a summary of notes and responses
transmitted for each notesfile with traffic.
.PP
The
.B -o
option specifies that only articles that arrived on the 
local system after
.I date-spec
are to be considered for transmittal to news.
The timestamp file is updated after the execution of 
.I newsoutput
whether the 
.B -o
option is specified.
.PP
The actual method of insertion into the news system is determined
in the ``net.how'' file by the transmission method for the
pseudo-site ``Usenet''.
Mappings between notesfiles and newsgroups are specified in the
file ``/usr/spool/notes/.utilities/newsgroups''.
.PP
The
.I newsinput
program accepts articles from the news system,
parses headers,
maps newsgroups to appropriate notesfiles,
and inserts the articles.
News' control messages are intercepted and placed in a 
maintainence notesfile for review by the system manager.
When a news article maps to a non-existent notesfile,
the notesfile is automatically created.
.PP
The
.B "-t"
directs
.I newsinput
to NOT log cases where the
.I "Message-Id"
is longer than the reserved space in the note descriptor.
These occurences are normally logged in the nfmaint notesfile.
The
.B "-1"
option directs
.I newsinput
to place the article in the first notesfile on the
.I "Newsgroups"
line.
The normal action is to place a copy of the article in each notesfile
specified on the
.I "Newsgroups"
line.
.PP
Typically
.I newsoutput
is run by cron(8) at intervals ranging from hourly to daily.
Newsinput is invoked automatically by the news system when
it propogates articles. 
See appendix B of 
.ul
The Notesfile Reference Manual
for a more detailed discussion on configuring and installing
a notesfile-news gateway.
.PP
The notesfile-news mapping function used the file
/usr/spool/notes/.utilities to map names between the two systems.
The mapping functions allow simple one-for-one name mapping,
one-to-many mapping,
and 
mapping ``basenotes'' and responses from one notesfile to 
separate newsgroups.
See appendix B of
.ul
The Notesfile Reference Manual
for more information on how this works.
.SH BUGS
News articles which can definitely be marked as responses
because they have a ``References'' line but whose parents
can't be found by article-ID or by guessing from the title
are not counted as orphans.
This is due to problems arising from mapping of article-ID's
to allow a notesfile and news system to coexist on the
same host.
.PP
newsinput doesn't correctly do one-to-many mapping of
newsgroups to notesfiles.  Newsoutput does do
one-to-many mapping of notesfiles to newsgroups.
.PP
Since the author's system may not have seen the article yet 
(if he wrote it via nfmail for example), the ``Path:'' header
line may not point anywhere near the author.
.PP
The ``Path'' line generated contains full-domain information
and doesn't necessarily point at the author.  News systems
2.10.2 and earlier still assume that the last component of
this header line is an author name and treat it specially.
.PP
The Message-Id field is set at 32 bytes in Notes 1.7.
This should be made longer.
64 bytes is probably adequate.
.PP
Because colons are used as delimiters in the notesfile
transfer format, Message-Id's with colons are declared
"UnManageable" and ignored.
.SH FILES
.PD 0
.TP 40
/usr/spool/notes/.utilities
where these programs live.
.TP 40
/usr/spool/notes/.utilities/newsgroups
mapping between notesfiles and newsgroups
.TP 40
/usr/spool/notes/.utilities/net.how
Pseudo-site ``Usenet'' specifies how to insert news
articles.
.PD
.SH SEE ALSO
news(1),
notes(1),
nfxmit(8),
uucp(1),
.br
.ul
The Notesfile Reference Manual
.SH AUTHORS
.nf
Ray Essick (uiucdcs!essick, essick%uiuc@csnet-relay.arpa)
Department of Computer Science
University of Illinois at Urbana-Champaign
Urbana, IL
.sp
Rob Kolstad (kolstad@convex.UUCP)
CONVEX Computer Corporation
Richardson, TX
.fi


usr/new/notes/src/newsoutput.c
#define	MAINLINE
#include "parms.h"
#include "structs.h"
#include "newsgate.h"
#include "net.h"
#ifdef	SYSLOG
# include <syslog.h>
# undef stderr
# define	fprintf		syslog
# define	stderr		LOG_NOTICE
#endif	SYSLOG

#ifdef	pdp11
#define	gatesysnames _gtsysn
#define	gatesyscount _gtsysc
#endif

#ifdef	RCSIDENT
static char rcsid[] = "$Header: /usr/local/src/usenet/notes/src/RCS/newsoutput.c,v 1.8 88/11/11 00:59:03 paul Exp $";
#endif	RCSIDENT


/*
 *	newsoutput - place the specified notesfiles out to the
 *	news susbsystem. 
 *
 *	Original Coding:	Ray Essick	April 1982
 *	Modified to better handle multiple system gateing
 *				Ray Essick	September 1982
 */

static char *whichsys;					/* who dumping for */
static int  verbosity = 0;				/* quiet */
static int  backward = 0;				/* do old format */
char    rnewscmd[CMDLEN] = DFLTRNEWS;			/* how to rnews */
int     sendclass = 0;					/* classes to send */
int     usetime = NORMSEQ;				/* user supplied */
int     gatesyscount = 0;				/* actives slots in */
char   *gatesysnames[GATEMAX];				/* list of names */

newsone (nfname)
char   *nfname;
{
    if (verbosity)
    {
	printf ("\t%s\n", nfname);
	fflush (stdout);
    }
    return newsout (nfname, backward, usetime, verbosity);
}


main (argc, argv)
char  **argv;
{
    int     i;
    struct nflist_f *nfptr;
    struct nethow_f how;
    struct when_f   ztime;
    char    fmtdate[DATELEN];				/* formatted date */
    int     aflag,
            cflag,
            sflag;

    startup (argc, argv);				/* common init */

    if (argc == 1)
    {
	printf ("Usage: %s [-ssitename | -a] [-v] [-b] [-f file] topic1 [topic2 ...]\n", argv[0]);
	exit (BAD);
    }

    if (globuid != Notesuid)
    {
	printf ("Sorry, only notes 'owner' can send notes to news\n");
	exit (BAD);
    }

#ifdef	SYSLOG
    openlog ("newsoutput", LOG_PID, SYSLOG);
#endif	SYSLOG
    whichsys = System;					/* defaults to local */
    aflag = cflag = sflag = 0;				/* none done */
    gatesyscount = 0;					/* none there */
    verbosity = 0;					/* no messages */
    backward = 0;					/* just new format */
    usetime = NORMSEQ;					/* non-seq time */
    strcpy (rnewscmd, DFLTRNEWS);			/* default rnews */
    getnet (RNEWSNAME, NULL, &how);			/* how to get there */
    if (how.nh_xmit != (char *) NULL)			/* specified? */
	strcpy (rnewscmd, how.nh_xmit);			/* load it */

    for (i = 1; i < argc; i++)
    {
	if (argv[i][0] == '-')
	    switch (argv[i][1])
	    {

		case 's': 				/* gateway for */
		    sflag = 1;
		    if (argv[i][2] != '\0')		/* -ssite */
		    {
			whichsys = &argv[i][2];
		    }
		    else
		    {					/* -s sitename */
			if (++i == argc)		/* no sitename */
			{
			    fprintf (stderr, "-s must be followed by a sitename\n");
			    exit (BAD);
			}
			whichsys = argv[i];		/* mark site */
		    }
		    gatesyscount = 1;			/* fill in table */
		    gatesysnames[0] = whichsys;
		    break;

		case 'a': 				/* all non-local */
		    sendclass |= NEWS_ALLSEND;
		    aflag = 1;
		    whichsys = NEWSSYS;			/* it's sequencer */
		    break;

		case 'c': 				/* file of systems */
		    cflag = 1;
		    whichsys = NEWSSYS;			/* use the global seq */
		    if (++i == argc)
		    {
			printf ("%s: -c option must be followed by filename\n", Invokedas);
			exit (BAD);
		    }
		    loadgate (argv[i]);			/* load the table */
		    break;

		case 'v': 				/* messages per group */
		    verbosity++;
		    break;

		case 'f': 				/* process a file */
		    if (++i == argc)			/* no filename */
		    {
			fprintf (stderr, "-f must be followed by a filename\n");
			exit (BAD);
		    }
		    readrc (argv[i]);			/* read the file */
		    break;

		case 'b': 				/* include old headers */
		    backward++;
		    break;

		case 't': 				/* explicit time */
		case 'o': 				/* compatibility */
		    if (++i == argc)
		    {
			fprintf (stderr, "-t option requires following date\n");
			exit (BAD);
		    }
		    switch (parsetime (argv[i], &ztime))
		    {
			case 0: 			/* ok */
			    usetime = BASESEQ;		/* use this time */
			    Basetime = ztime;		/* store it */
			    sprdate (&ztime, fmtdate);	/* format */
			    printf ("%s: Sending articles since %s\n",
				    Invokedas, fmtdate);
			    break;
			case -1: 			/* no good */
			    fprintf (stderr, "%s: unable to parse time `%s'\n",
				    Invokedas, argv[i]);
			    exit (BAD);
			case -2: 			/* in future */
			    fprintf (stderr, "%s: parsed date (%s) is in the future\n",
				    Invokedas, argv[i]);
			    exit (BAD);
		    }
		    break;


		default: 
		    printf ("Bad switch: %c\n", argv[i][1]);
		    exit (BAD);
	    }
	else
	    expand (argv[i]);				/* put it in the list */
    }

/*
 *	now, run down the list and send each notesfile.
 *	check to make sure some things are used consistently first.
 */

    if (aflag + cflag + sflag > 1)
    {							/* -a and -s */
	fprintf (stderr, "%s: -a, -c and -s mutually exclusive\n", Invokedas);
	exit (BAD);
    }
    if (gatesyscount == 0 && !cflag)			/* default to local */
    {
	/* 
	 * check against cflag allows empty -c files to "work"
	 */
	gatesysnames[0] = whichsys;
	gatesyscount = 1;				/* just us */
    }
    if (1)						/* was verbosity */
    {							/* now always... */
	printf ("%s: Sending news articles through command: %s\n",
		Invokedas, rnewscmd);
	if (sendclass & NEWS_ALLSEND)
	{
	    printf ("Sending to News for any system\n");
	}
	else
	{
	    int     col;				/* count columns */

	    /* 
	     * WE WANT TO SORT THE LIST AT THIS POINT
	     * so that the cansend() routine in newsout.c can do a binary
	     * search to make things faster.
	     */
	    printf ("Gateway for the following %d system(s):\n", gatesyscount);
	    for (i = 0, col = 0; i < gatesyscount; i++)
	    {
		printf (" %16s", gatesysnames[i]);
		if (++col % 4 == 0)
		    putc ('\n', stdout);		/* multi-column */
	    }
	}
	printf ("\n================\n");
    }
    fflush (stdout);					/* make sure in order */
    sprintf (Seqname, "%s:%s", NEWSSEQ, whichsys);
    while ((nfptr = nextgroup ()) != (struct nflist_f *) NULL)
	newsone (nfptr -> nf_name);

    exit (GOOD);					/* all done */
}

/*
 *	loadgate(filename)
 *
 *	fill in the system table.
 */

loadgate (filename)
char   *filename;
{
    FILE * fptr;
    char    onesystem[256];				/* system name */

    if ((fptr = fopen (filename, "r")) == (FILE *) NULL)
	return (-1);					/* bah humbug */

    while (fscanf (fptr, "%s", onesystem) == 1)
    {
	gatesysnames[gatesyscount++] = strsave (onesystem);
    }

    fclose (onesystem);
    return (0);
}
